### PR TITLE
ENH: Implement GPU version of spectral decomposition as linear regression fallback

### PR DESCRIPTION
**Pull request by david-cortes-intel**
## Description

As a follow up from https://github.com/oneapi-src/oneDAL/pull/2930

This PR adds a GPU version of the fallback for linear regression based on spectral decomposition, which mirrors the CPU version. It reuses the same tests, but this time executing them with fp32 too.

A few things I wasn't sure about:
* Here I'm taking the minimum of the diagonal of the Cholesky factorization using a sycl reduction class, but I'm not sure if that's the usual way in oneDAL.
* I am also not 100% sure if it is guaranteed that this operation (minimum of the diagonal) with an 'at_device' at the end would force execution to wait until the operation is completed before returning the value.
* There's an operation here for which oneMKL doesn't provide the equivalent LAPACK functionality, which is: "column-major matrix divided by a broadcasted row vector" (i.e. multiply each row of the matrix by a scalar from the vector). I implemented it as a 2D for-loop with a multiplication by the inverse (which is faster but less precise), but I don't know if this is better than a 1D loop that would take a whole row, as rows are contiguous in memory.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

Not applicable (addition is a fallback for cases that would have failed previously).

### CODE REVIEW

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:95]: Does it make sense to replace ndview and array with ndarray?

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:212]: Can it be potentially reused in scenario with more than 1 event? Should it be extended with const bk::event_vector& deps

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:103]: It requires finding the index of the first value in a sorted array that is above a certain threshold. I couldn't think of any fast way of doing it on GPU which wouldn't involve just transferring the data to CPU, either at once or one-by-one.

Do you have a suggestion for some better way of doing it?

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:241]: As I remember we use try catch only in tests/examples. Does it make sense to avoid it here?

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:212]: It should be theoretically possible, but the function right now is only used in this file.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:241]: It's meant to catch an exception thrown by oneMKL in the call to `potrf`, over which we have no control AFAIK. An alternative would be to modify the wrapper, but it's also used elsewhere in places where the exception is not catched (so a user error is thrown).

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:95]: There is magic conversion that works:
https://github.com/oneapi-src/oneDAL/blob/main/cpp/oneapi/dal/algo/pca/backend/gpu/misc.hpp#L77
here ndarray has been provided in syevd

Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:95]: Thanks. And what would be the difference between `array` and `ndarray`? Looks like they use the same allocator.

**Comment by david-cortes-intel** [comment]
Another thing that I'm now not very sure about is: should calls to oneMKL lapack manually call `queue.wait_and_throw()` in order to catch runtime exceptions from them?

**Comment by david-cortes-intel** [comment]
After some investigation regarding the `diagonal_minimum` function, turns out the issue with the function for diagonal minimum was that it wasn't initializing the reducer object, despite what the docs mention:
https://github.khronos.org/SYCL_Reference/iface/reduction-variables.html#known-identities

Adding a manual initialization did the trick.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:95]: In general would be no difference, but you can avoid allocation of ndview

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:103]: no, sounds good, thanks

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp]: Changed to ndarray.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Vika-F) [inline cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:111]: Please add new error message into `error_handling.hpp` and use it like it is done in other algorithms:
https://github.com/oneapi-src/oneDAL/blob/main/cpp/oneapi/dal/algo/knn/backend/gpu/train_kernel_brute_force_dpc.cpp#L38

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Vika-F) [inline cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:146]: Please use existing wrappers to oneMKL:
https://github.com/oneapi-src/oneDAL/blob/main/cpp/oneapi/dal/backend/primitives/blas/gemv.hpp

Or add a new wrapper if it is needed.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Vika-F) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:245]: I am Ok with having a `try` in the code; but let's make it more clear for the readers - please add the description of what is happening here - why the system is tries to be solved twice in some cases.

**Review by Vika-F — COMMENTED** [review]
The overall changes look good to me, they just need to be aligned a bit with the codebase. Please see the comments.

**Comment by Vika-F** [comment]
> Another thing that I'm now not very sure about is: should calls to oneMKL lapack manually call `queue.wait_and_throw()` in order to catch runtime exceptions from them?

@david-cortes-intel Good point. I think yes, `.wait_and_throw()` is needed after the last oneMKL function call in `try` to catch the exception.
Otherwise it can be missed as oneMKL calls run asynchronously on device.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp]: Changed comments format to three backslashes and added a comment about the expected dimensions of the input arrays.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:199]: I think it should get overwritten by passing `sycl::property::reduction::initialize_to_identity()`: https://github.khronos.org/SYCL_Reference/iface/reduction-variables.html#id2

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:232]: Added a comment about the logic for the try-catch block.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp]: Moved the error message to the file with internal error messages.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp]: Changed them to the internal wrappers for `gemm` and `gemv`, but I'd say their logic is quite hard to follow.

**Comment by david-cortes-intel** [comment]
> > Another thing that I'm now not very sure about is: should calls to oneMKL lapack manually call `queue.wait_and_throw()` in order to catch runtime exceptions from them?
> 
> @david-cortes-intel Good point. I think yes, `.wait_and_throw()` is needed after the last oneMKL function call in `try` to catch the exception. Otherwise it can be missed as oneMKL calls run asynchronously on device.

Thanks for confirming. Added a manual call to `wait_and_throw`.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Vika-F) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:199]: It looks like it shouldn't. It gets overwritten only for integral data types. Please take a look at the table here:
https://github.khronos.org/SYCL_Reference/iface/reduction-variables.html

Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:199]: For `sycl::minimum`, it mentions under "Available Only When":
```
std::is_floating_point_v<AccumulatorT> || std::is_same_v<std::remove_cv_t<AccumulatorT>, sycl::half>
```

Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:199]: It has two rows in the table for `sycl::minimum` BTW.

File: cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp
Comment (Vika-F) [inline cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp:237]: That should be the last portion of comments.

1. Please provide some descriptive name for 1e-6 threshold. We are trying to reduce the number of magic constants in our code.

2. Throwing oneMKL exception here looks like a hack to call the code in catch.
It would be better just to call spectral decomposition here and in catch. Maybe re-write it a bit in order not to duplicate the code with array copying.

Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp]: Added a comment about the threshold and changed the logic of the fallback to a named section.

Comment (Vika-F) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp]: Maybe I do not understand something, but why this block cannot be just a function? For the sake of simplicity.

`goto` in DPC++ code looks pretty weird. And besides the performance aspect, the use of `goto` make the code harder to understand. As the execution order and the call stack can be very hard to track.

Comment (david-cortes-intel) [reply cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp]: Changed it into a function.

**Review by Vika-F — APPROVED** [review]
Let's wait for a [semi-]green CI and LGTM.

**Pull request merged**
